Add Unicode Support

adam-grant-hendry opened the pull request:
Adds unicode support by allowing configurable encodings to be specified (defaults to `utf-8`).

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `./scripts/format` and `./scripts/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes

Fixes: Issue #516

adam-grant-hendry commented:
@Lee-W @woile Please review when you get a chance. Thanks!

Lee-W commented on .github/workflows/pythonpackage.yml:
```diff
@@ -7,7 +7,7 @@ jobs:
       matrix:
         python-version: ["3.6", "3.7", "3.8", "3.9", "3.10"]
-        platform: [ubuntu-latest, macos-latest, windows-latest]
+        platform: [ubuntu-20.04, macos-latest, windows-latest]
```
Is there any reason we need to pin it to a certain version?

adam-grant-hendry replied:
No worries and thanks for reviewing! The CI workflow `actions/setup-python@v4` kept failing on `ubuntu-latest` with:

`Error: Version 3.6 with arch x64 not found`

This error was reported in

https://github.com/actions/setup-python/issues/544 

and the recommended fix is to hard code to Ubuntu 20.04 if Python 3.6 is needed since they no longer are supporting Python 3.6.

I figured this could be acceptable until we sunset Python 3.6.

Lee-W replied:
ah, got it. Thanks for replying! I'm now encountering the same issue on my own PR haha

adam-grant-hendry replied:
Ya, it just started happening. The team didn’t make tarballs of 3.6 for the new distro and chose not to add them since 3.6 is already deprecated.

Side note, it’s usually best to specify specific versions rather than latest for this reason. You get more repeatable results that way, so it might be good to do the same for the other OS’s.

Lee-W reviewed (approved):
@adam-grant-hendry Thanks for the update! Just one minor discussion needed on my side.

@woile I'm planning on merging this next week. Let me know if you need more time to take a deeper look. Thanks!

woile commented on commitizen/defaults.py:
```diff
@@ -65,6 +67,7 @@ class Settings(TypedDict, total=False):
     "update_changelog_on_bump": False,
     "use_shortcuts": False,
     "major_version_zero": False,
+    "encoding": "utf-8",
```
shouldn't this be

```py
"encoding": encoding,
```

adamgranthendry replied:
Good point. Should we also change

```python
"name": "cz_conventional_commits"
```

to

```python
"name": name
```

adam-grant-hendry replied:
Good point. Should we also change

```python
"name": "cz_conventional_commits"
```

to

```python
"name": name
```

woile commented:
LGTM, the only thing missing is documentation: https://commitizen-tools.github.io/commitizen/config/

Ideally, I think the `cz init` should ask for the encoding **if** you are on windows, this would make windows users aware of potential encoding issues

Lee-W commented:
It seems CI failed on windows 🤔

adam-grant-hendry commented:
@Lee-W I didn't push any commits yesterday, but it is showing that I did through `Unverified` commits. By chance, was this you?

Lee-W commented:
Hi @adam-grant-hendry , yes I tried to rebase the master branch to this one to see if the latest update fix this branch

woile commented:
Is this still valid? Please rebase 🙏🏻

adam-grant-hendry commented:
@woile @Lee-W Apologies for taking so long: I've been away for quite some time. My gpg signature keys are outdated, but the commits are by me.

`mypy` is still unhappy with line 6 of `out.py` (See: https://github.com/python/typeshed/issues/3049). I added a `#type: ignore`, but Python 3.7 reads this as an unused type ignore (See: https://github.com/commitizen-tools/commitizen/actions/runs/5491623964/jobs/10008334198#step:5:251), so I'm stuck.

Only remaining item is to add documentation. I added `encoding` as a keyword argument to `changelog.py/get_metadata` so as to not break backwards-compatibility with the added `scheme` argument.

adam-grant-hendry commented:
@woile @Lee-W Everything is good now. Could you please re-review the changes and make sure I added everything you requested?

Lee-W commented:
@adam-grant-hendry Sure! I'll take a look these days.

Lee-W commented on commitizen/bump.py:
```diff
@@ -45,7 +45,12 @@ def find_increment(
+    current_version: str,
+    new_version: str,
+    files: list[str],
+    encoding: str,
+    *,
```
If possible, I'd like to keep newly added argument as keyword only argument
```suggestion
    *,
    encoding: str,
```

adam-grant-hendry replied:
Resolved in commit `e9d73c5f`.

Lee-W commented on commitizen/changelog.py:
```diff
@@ -213,7 +213,9 @@ def parse_title_type_of_line(value: str) -> str | None:
 
-def get_metadata(filepath: str, scheme: VersionScheme = Pep440) -> dict:
+def get_metadata(
+    filepath: str, scheme: VersionScheme = Pep440, encoding: str = "utf-8"
```
Is the default `"utf-8"` required here? if so, should we add this default elsewhere as well?

adam-grant-hendry replied:
Since `scheme` is a keyword parameter, a non-keyword parameter cannot come after it in the parameter list. I added `"utf-8"` so as not reorder the parameters and break backwards compatibility.

Instead, I can import `defaults` and use `encoding` defined there:

```python
from commitizen.defaults import encoding
...

def get_metadata(
    filepath: str, scheme: VersionScheme = Pep440, encoding: str = encoding
) -> dict:
    ...
```

Python still [defaults to the 8-bit 1252 character set encoding on Windows](https://docs.python.org/3/using/windows.html#utf-8-mode). Making `"utf-8"` the default encoding is a good choice and shouldn't be problematic (see `Code that may break` in [PEP 528](https://peps.python.org/pep-0528/) and [PEP 529](https://peps.python.org/pep-0529/) for impetus).

Files opened in binary mode [return contents as `bytes` objects without any decoding](https://docs.python.org/3/library/functions.html#open), so `encoding` wasn't added to those uses of `open` (nor for trivial cases like `commands/init.py/is_python_poetry()`.

I'll add `.encoding` everywhere else where it was missed.

adam-grant-hendry replied:
Resolved in commit `e9d73c5f`.

Lee-W reviewed (commented):
Hi @adam-grant-hendry , sorry for the late review. Overall the changes are good, but I think we should drop commit `3430edba7a5b1b4f59e5b7aa48e03bd15ccc32a4` and `9af6c748eb1084c5239e8d8058e7eddd59abbfb5` which downgrades the version of our GitHub actions workflow. Also, I notice some `open` and `smart_open` miss this `encoding` feature. Is this intentional? Or just accidentally missed? Thanks!

adam-grant-hendry commented:
> Hi @adam-grant-hendry , sorry for the late review. Overall the changes are good, but I think we should drop commit `3430edba7a5b1b4f59e5b7aa48e03bd15ccc32a4` and `9af6c748eb1084c5239e8d8058e7eddd59abbfb5` which downgrades the version of our GitHub actions workflow. Also, I notice some `open` and `smart_open` miss this `encoding` feature. Is this intentional? Or just accidentally missed? Thanks!

No worries at all. Thank you for your review!

1. Yes, good find: we should definitely drop those commits then
2. That was certainly an accidental miss on my part. I’ll review, make changes, and push up the fixes

I probably can’t get to this today, but I can start tomorrow. I’ll see if I can request your re-review by Monday morning.

Thanks again!

adam-grant-hendry commented:
@Lee-W I made all the changes discussed. Please re-review this PR at your earliest convenience. Thank you!

Lee-W commented:
@adam-grant-hendry Sure! I'll try to take a look this weekend

Lee-W reviewed (approved):
Overall, I'm good with this PR. But we missed a few places missed. Wondering is it designed this way or just missed. https://github.com/commitizen-tools/commitizen/blob/114501de6c9a0a763b291f2ec3b3251522049d94/commitizen/commands/init.py#L38, https://github.com/commitizen-tools/commitizen/blob/114501de6c9a0a763b291f2ec3b3251522049d94/hooks/prepare-commit-msg.py#L63.

But, yep looks like they might be something not affected 🤔

I'm planning on merging this next week. @woile Please let me know if you want to take a deeper looks 🙂

Lee-W commented:
@adam-grant-hendry ah, just notice a conflict. might need your help to resolve it.

adam-grant-hendry commented:
> ...But we missed a few places missed. Wondering is it designed this way or just missed.
> 
> https://github.com/commitizen-tools/commitizen/blob/114501de6c9a0a763b291f2ec3b3251522049d94/commitizen/commands/init.py#L38
> 
> ,
> https://github.com/commitizen-tools/commitizen/blob/114501de6c9a0a763b291f2ec3b3251522049d94/hooks/prepare-commit-msg.py#L63
> 
> .
> But, yep looks like they might be something not affected 🤔
> 
I purposefully left these out as I considered them "trivial" cases. By that I mean the `"pyproject.toml"` case is only looking for the `"[tool.poetry]"` table heading, which should generally not be affected by whether the file is opened in `utf-8` mode or not. As for `"/dev/tty"`, there is no equivalent to `"/dev/tty"` on Windows machines, so this would only run properly from a Linux environment, in which case the encoding is already `utf-8` by default.

Lee-W commented:
Sounds good 👍 Just want to confirm it. I think we're good to merge it after resolving the conflict. Thanks!

woile reviewed (approved):
LGTM as well. Thanks!

adam-grant-hendry commented on commitizen/bump.py:
```diff
@@ -45,7 +45,12 @@ def find_increment(
+    new_version: str,
+    files: list[str],
+    *,
+    check_consistency=False,
```
Resolved in `69c90`

adam-grant-hendry commented:
> @adam-grant-hendry ah, just notice a conflict. might need your help to resolve it.

I had to pull in the latest commits to `master` and rebase. It's all good to go now!

Lee-W commented:
Many thanks! Just merged
